pkg/apis/pingcap/v1alpha1/tidbmonitor_types.go

```diff
@@ -0,0 +1,133 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// +k8s:openapi-gen=true
+// TidbMonitor encode the spec and status of the monitoring component of a TiDB cluster
+type TidbMonitor struct {
+	metav1.TypeMeta `json:",inline"`
+	// +k8s:openapi-gen=false
+	metav1.ObjectMeta `json:"metadata"`
+
+	// Spec defines the desired state of TidbMonitor
+	Spec TidbMonitorSpec `json:"spec"`
+
+	// +k8s:openapi-gen=false
+	// Most recently observed status of the TidbMonitor
+	Status TidbMonitorStatus `json:"status"`
+}
+
+// +k8s:openapi-gen=true
+// TidbMonitor spec encode the desired state of tidb monitoring component
+type TidbMonitorSpec struct {
+	Clusters        []TidbClusterRef  `json:"clusters"`
+	Prometheus      PrometheusSpec    `json:"prometheus"`
+	Grafana         GrafanaSpec       `json:"grafana"`
+	Reloader        ReloaderSpec      `json:"reloader"`
+	Initializer     InitializerSpec   `json:"initializer"`
+	ImagePullPolicy corev1.PullPolicy `json:"imagePullPolicy,omitempty"`
+	Persistent      bool              `json:"persistent,omitempty"`
+	// +optional
+	StorageClassName string `json:"storageClassName,omitempty"`
+	// +optional
+	Storage string `json:"storage,omitempty"`
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+// PrometheusSpec is the desired state of prometheus
+type PrometheusSpec struct {
+	MonitorContainer `json:",inline"`
+
+	LogLevel    string      `json:"logLevel,omitempty"`
+	ReserveDays int         `json:"reserveDays,omitempty"`
+	Service     ServiceSpec `json:"service,omitempty"`
+}
+
+// GrafanaSpec is the desired state of grafana
+type GrafanaSpec struct {
+	MonitorContainer `json:",inline"`
+
+	LogLevel string      `json:"logLevel,omitempty"`
+	Service  ServiceSpec `json:"service,omitempty"`
+	Username string      `json:"username,omitempty"`
+	Password string      `json:"password,omitempty"`
+	// +optional
+	Envs map[string]string `json:"envs,omitempty"`
+}
+
+// ReloaderSpec is the desired state of reloader
+type ReloaderSpec struct {
+	MonitorContainer `json:",inline"`
+	Service          ServiceSpec `json:"service,omitempty"`
+}
+
+// InitializerSpec is the desired state of initializer
+type InitializerSpec struct {
+	MonitorContainer `json:",inline"`
+	// +optional
+	Envs map[string]string `json:"envs,omitempty"`
```

Comment:
We could set K8S_PROMETHEUS_URL via Envs? Grafana nodesInfo dashboard relies on this field.

Reply:
Yes we could